### PR TITLE
kernel/binary_manager: Release all kernel semaphores taken by all threads in binary when binary recovery

### PR DESCRIPTION
**Pull request by jeongarmy**
- Manage the list of semaphore holder with enabling binary manager recovery
- Release all kernel semaphores taken by all task/pthreads when binary is reloaded by searching a list

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [db43263c63af06219267963807216fc01d7617a4] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [db43263c63af06219267963807216fc01d7617a4] - Code Rule Check OK.

File: lib/libc/semaphore/sem_init.c [hunk 1]
```diff
@@ -60,11 +60,41 @@
...
+#if defined(CONFIG_BINMGR_RECOVERY) && defined(__KERNEL__)
+sq_queue_t g_sem_list;
+
+static void sem_register(FAR sem_t *sem)
+{
+	sem_t *sem_ptr;
+	irqstate_t flags;
+
+	flags = irqsave();
+	if (!sq_empty(&g_sem_list)) {
```
Comment (jeongchanKim) [inline lib/libc/semaphore/sem_init.c]: I think this if condition is not needed, because if g_sem_list is NULL, then below sem_ptr will be NULL.

Comment (sijupunnnoose) [inline lib/libc/semaphore/sem_init.c]: You dont need this check here, while() will take care of it.

File: os/kernel/semaphore/sem_post.c [hunk 2]
```diff
@@ -174,8 +177,10 @@ int sem_post(FAR sem_t *sem)
 			for (stcb = (FAR struct tcb_s *)g_waitingforsemaphore.head; (stcb && stcb->waitsem != sem); stcb = stcb->flink) ;
 
 			if (stcb) {
-				sem_addholder_tcb(stcb, sem);
-
+				if (!(sem->flags & FLAGS_ZEROSEM)) {
```
Comment (jeongchanKim) [inline os/kernel/semaphore/sem_post.c]: You already made `IS_ZERO_SEM` macro.
How about using this macro?

File: os/kernel/semaphore/semaphore.h [hunk 3]
```diff
@@ -107,14 +107,16 @@ void sem_recover(FAR struct tcb_s *tcb);
  * holders of semaphores.
  */
 
-#ifdef CONFIG_PRIORITY_INHERITANCE
+#if defined(CONFIG_PRIORITY_INHERITANCE) || defined(CONFIG_BINMGR_RECOVERY)
```
Comment (jeongchanKim) [inline os/kernel/semaphore/semaphore.h]: You can change this line to `#if SAVE_SEM_HOLDER`

**Comment by seinfra** [comment]
Target : [df20f4021ecd5795c22af8cea4498cfeabbd62a7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [df20f4021ecd5795c22af8cea4498cfeabbd62a7] - Code Rule Check OK.

File: os/include/semaphore.h [hunk 4]
```diff
@@ -134,12 +149,20 @@ typedef struct sem_s sem_t;
...
  */
-#ifdef CONFIG_PRIORITY_INHERITANCE
+#if SAVE_SEM_HOLDER
+#ifdef CONFIG_BINMGR_RECOVERY
+#if CONFIG_SEM_PREALLOCHOLDERS > 0
+#define SEM_INITIALIZER(c) {NULL, (c), FLAGS_INITIALIZED, NULL} /* flink, semcount, flags, hhead */
+#else
+#define SEM_INITIALIZER(c) {NULL, (c), FLAGS_INITIALIZED, SEMHOLDER_INITIALIZER} /* flink, semcount, flags, holder */
+#endif
+#else // CONFIG_APP_BINARY_SEPARATION
```
Comment (jeongchanKim) [inline os/include/semaphore.h]: This else is not for CONFIG_APP_BINARY_SEPARATION.

File: os/kernel/binary_manager/binary_manager_recovery.c [hunk 5]
```diff
@@ -62,6 +63,38 @@ static void binary_manager_board_reset(void)
...
+#if CONFIG_SEM_PREALLOCHOLDERS > 0
+		for (holder = sem->hhead; holder; holder = holder->flink)
+#else
+		holder = &sem->holder;
+#endif
+		{
+			if (holder && holder->htcb && holder->htcb->group && holder->htcb->group->tg_loadtask == binid) {
+				/* It doesn't release if the semaphore with zero value and semcount is larger than or equal with 0. */
+				if (IS_ZERO_SEM(sem) && sem->semcount >= 0) {
+					sem = sq_next(sem);
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_recovery.c]: In this case, sem will be changed to next, but holder is not.
Please check this logic.

**Comment by seinfra** [comment]
Target : [fcf9dacd889d772557e003c3fce87b2313aaab8f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fcf9dacd889d772557e003c3fce87b2313aaab8f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [6ec750ce1d53db1c9854329fe6a178a521249212] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6ec750ce1d53db1c9854329fe6a178a521249212] - Code Rule Check OK.

File: os/include/semaphore.h [hunk 6]
```diff
@@ -79,13 +79,21 @@
 #define PRIOINHERIT_FLAGS_DISABLE (1 << 0) /* Bit 0: Priority inheritance
 					    * is disabled for this semaphore */
 #define FLAGS_INITIALIZED         (1 << 1) /* Bit 1: This semaphore initialized */
+#ifdef CONFIG_BINMGR_RECOVERY
+#define FLAGS_ZEROSEM             (1 << 2) /* Bit 2: The semaphore for signaling */
```
Comment (jeongchanKim) [inline os/include/semaphore.h]: ZEROSEM is not good for understanding. How about using SIG_SEM or another?

File: os/include/semaphore.h [hunk 7]
```diff
@@ -79,13 +79,21 @@
...
+#endif
+
+/* Save semaphore holder data with enabling priority inheritance or binary manager recovery */
```
Comment (jeongchanKim) [inline os/include/semaphore.h]: I think
`Save semaphore holder data when priority inheritance or binary manager recovery is enabled.`
is better.

File: os/kernel/semaphore/sem_post.c [hunk 8]
```diff
@@ -140,9 +143,9 @@ int sem_post(FAR sem_t *sem)
 		saved_state = irqsave();
 
 		/* Perform the semaphore unlock operation. */
-
+		htcb = this_task();
```
Comment (jeongchanKim) [inline os/kernel/semaphore/sem_post.c]: I think we can use stcb in here.
And if we use htcb, we don't need to use rtcb when CONFIG_SEMAPHORE_HISTORY is enabled.

File: os/kernel/binary_manager/binary_manager_recovery.c (hunk 5 above)
Comment (jeongarmy) [reply os/kernel/binary_manager/binary_manager_recovery.c]: right. I modified.

File: os/kernel/semaphore/sem_destroy.c [hunk 9]
```diff
@@ -85,6 +88,27 @@
...
+{
+	sem_t *sem_ptr;
+	irqstate_t flags;
+
+	flags = irqsave();
+	sem_ptr = (sem_t *)sq_peek(&g_sem_list);
+	while (sem_ptr) {
+		if (sem_ptr == sem) {
+			/* Remove semaphore from a list of kernel semaphore */
+			sq_rem((FAR sq_entry_t *)sem, &g_sem_list);
```
Comment (jeongchanKim) [inline os/kernel/semaphore/sem_destroy.c]: sq_rem searches the correct sem.
So we don't need to run the while loop for searching.

**Comment by seinfra** [comment]
Target : [ac7c102326ff49332f833ff91f4e47256231a51e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ac7c102326ff49332f833ff91f4e47256231a51e] - Code Rule Check OK.

File: lib/libc/semaphore/sem_init.c [hunk 10]
```diff
@@ -60,11 +60,39 @@
...
+#endif
 
 #ifdef CONFIG_SEMAPHORE_HISTORY
 #include <tinyara/debug/sysdbg.h>
 #endif
 
+#if defined(CONFIG_BINMGR_RECOVERY) && defined(__KERNEL__)
+sq_queue_t g_sem_list;
+
+static void sem_register(FAR sem_t *sem)
```
Comment (jeongchanKim) [inline lib/libc/semaphore/sem_init.c]: How about separating register and unregister to another file? Because, sem_init is in lib, but sem_register is only for kernel.

**Comment by seinfra** [comment]
Target : [79d38755a2c954359ab17795ff3fab7089bd90d8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [79d38755a2c954359ab17795ff3fab7089bd90d8] - Code Rule Check OK.

File: os/include/semaphore.h [hunk 11]
```diff
@@ -248,6 +271,21 @@ int sem_unlink(FAR const char *name);
  */
 #endif
 
+#ifdef CONFIG_BINMGR_RECOVERY
+/**
+ * @cond
+ * @internal
+ */
+void sem_register(FAR sem_t *sem);
```
Comment (jeongchanKim) [inline os/include/semaphore.h]: You'd better move `sem_register` and `sem_unregister`  to tinyara/semaphore.h

File: os/kernel/semaphore/Make.defs [hunk 12]
```diff
@@ -1,6 +1,6 @@
 ###########################################################################
 #
-# Copyright 2016 Samsung Electronics All Rights Reserved.
+# Copyright 2019 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline os/kernel/semaphore/Make.defs]: Why did you change the license year? Is it needed?

File: os/kernel/semaphore/Make.defs [hunk 13]
```diff
@@ -57,6 +57,11 @@ CSRCS += sem_post.c sem_recover.c sem_reset.c sem_waitirq.c sem_tickwait.c
 
 ifeq ($(CONFIG_PRIORITY_INHERITANCE),y)
 CSRCS += sem_initialize.c sem_holder.c sem_setprotocol.c
+ifeq ($(CONFIG_BINMGR_RECOVERY),y)
+CSRCS += sem_list.c
+endif
```
Comment (jeongchanKim) [inline os/kernel/semaphore/Make.defs]: Please check that there is no missing for endif

File: os/kernel/semaphore/sem_list.c [hunk 14]
```diff
@@ -0,0 +1,119 @@
+/****************************************************************************
+ *
+ * Copyright 2016-2017 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline os/kernel/semaphore/sem_list.c]: Please check the license year.

File: os/kernel/semaphore/sem_list.c [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
```
Comment (jeongchanKim) [inline os/kernel/semaphore/sem_list.c]: This is not needed.

File: os/kernel/semaphore/sem_list.c [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+#include <semaphore.h>
+#include <errno.h>
```
Comment (jeongchanKim) [inline os/kernel/semaphore/sem_list.c]: `errno.h` is not needed.

**Comment by seinfra** [comment]
Target : [fe1e7ab29b8b25b726a15bdd681c70c32a484603] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fe1e7ab29b8b25b726a15bdd681c70c32a484603] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [bb375ebe355730c89bf18157a4ff335816441e58] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bb375ebe355730c89bf18157a4ff335816441e58] - Code Rule Check OK.

File: os/include/semaphore.h [hunk 17]
```diff
@@ -104,21 +110,26 @@ struct semholder_s {
 #else
 #define SEMHOLDER_INITIALIZER {NULL, 0}
 #endif
-#endif							/* CONFIG_PRIORITY_INHERITANCE */
+#endif							/* SAVE_SEM_HOLDER */
+
+#define IS_SIG_SEM(sem) ((sem->flags & FLAGS_SIGSEM) != 0)
```
Comment (jeongchanKim) [inline os/include/semaphore.h]: You'd better move this also to `tinyara/include.h`

File: os/kernel/binary_manager/binary_manager_recovery.c [hunk 18]
```diff
@@ -62,6 +63,42 @@ static void binary_manager_board_reset(void)
 #endif
 }
 
+static void recovery_release_binary_sem(int binid)
+{
+	sem_t *sem;
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_recovery.c]: For sem_t, you'd better include `semaphore.h`.

**Comment by seinfra** [comment]
Target : [bc0bd9b83a975e47b234ac492f835f425dce2586] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bc0bd9b83a975e47b234ac492f835f425dce2586] - Code Rule Check OK.

File: os/include/semaphore.h [hunk 19]
```diff
@@ -79,13 +79,19 @@
 #define PRIOINHERIT_FLAGS_DISABLE (1 << 0) /* Bit 0: Priority inheritance
 					    * is disabled for this semaphore */
 #define FLAGS_INITIALIZED         (1 << 1) /* Bit 1: This semaphore initialized */
+#define FLAGS_SIGSEM              (1 << 2) /* Bit 2: The semaphore for signaling */
```
Comment (jeongchanKim) [inline os/include/semaphore.h]: Setting this flag is only when `#if defined(CONFIG_BINMGR_RECOVERY) && defined(__KERNEL__)`

**Comment by seinfra** [comment]
Target : [95729d0a57f7d432f8470f8c0311b1016e176971] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [95729d0a57f7d432f8470f8c0311b1016e176971] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [adcad0111176e21c7728e6428594d7c1abb35d60] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [adcad0111176e21c7728e6428594d7c1abb35d60] - Code Rule Check OK.
